design/0000-launcher.md

```diff
@@ -0,0 +1,121 @@
+# Launcher
+
+## What is it?
+
+The Go Launcher (name TBD) is a configuration layer that chooses default values for configuration options that many OpenTelemetry users would ultimately configure manually, allowing for minimal code to quickly instrument with OpenTelemetry.
+
+## Motivation
+
+The goal of this Launcher is to help users that aren't familiar with OpenTelemetry quickly ramp up on what they need to get going and instrument their applications. [There is currently a lot of boilerplate code required to use OpenTelemetry](https://opentelemetry.io/docs/instrumentation/go/manual/#initializing-a-new-tracer), and users must make decisions early on before instrumentation works - for example, they must determine and find the appropriate exporters, set resource attributes, etc. This launcher is intended to simplify the setup and minimize decision-making overhead to be able to quickly get started.
+
+Separately, vendors may choose to provide a vendor-specific package compatible with this launcher for easier exporting to their own backend. Neither the launcher nor the vendor-specific package are required to use OpenTelemetry, and using the launcher does not require the end user to use a vendor-specific package.
+
+## Explanation
+
+The Go Launcher provides a single function to reduce the amount of code that needs to be written to get started, and provides additional, straightforward options to add in that still make additional configuration easy.
+
+The vendor-neutral launcher will live in the opentelemetry-go-contrib repo, and vendors may create separately hosted vendor-specific configuration packages that can be used with the launcher.
+
+### Getting started
+
+```shell
+go get github.com/opentelemetry-go-contrib/otel-launcher-go/launcher
+```
+
+### Configure
+
+Minimal setup - by default will send all telemetry to localhost:4317, with the ability to set environment variables for endpoint, headers, resource attributes, and more as listed in the configuration options noted below:
+
+```go
+import "github.com/opentelemetry-go-contrib/otel-launcher-go/launcher"
+
+func main() {
+    lnchr, err := launcher.ConfigureOpentelemetry()
+    defer lnchr.Shutdown()
+}
+```
+
+Or set headers directly in code instead:
+
+```go
+import "github.com/opentelemetry-go-contrib/otel-launcher-go/launcher"
+
+func main() {
+    lnchr, err := launcher.ConfigureOpentelemetry(
+        launcher.WithServiceName("service-name"),
+        launcher.WithHeaders(map[string]string{
+            "service-auth-key": "value",
+            "service-useful-field": "testing",
+        }),
+    )
+    defer lnchr.Shutdown()
+}
+```
+
+### Configuration Options
+
+| Config Option               | Env Variable                        | Required | Default              |
```

Comment:
I don't see any options related to the `BatchSpanProcessor`.  I'd assume that would be used and the user might need a way to modify it.  

This makes me wonder more generally what the transition path would be for a user who outgrows the launcher and wants to convert to more explicit initialization.  It would be awesome (though maybe overkill) if there was a `lnchr.Generate()` method that could emit the code that would be necessary to explicitly instantiate the launcher.

Reply:
The launcher uses the `BatchSpanProcessor` by default. There currently is not a plan to swap out the batch processor with, for example, the `SimpleSpanProcessor`. In our other distros (Java, .NET) we have been using the batch processor with no option to swap it out, and have not yet received any feedback requesting to make it configurable.

The `Generate()` function you mention does sound like a great stretch goal, but it may be too much for the initial implemenetation here.

Reply:
I certainly wouldn't suggest using the `SimpleSpanProcessor`, but the `BatchSpanProcessor` does have several configuration options.  We should consider where the breakpoint should be that forces users to drop the simple setup path and use fully manual SDK configuration.  I think that adjusting batch size or timing should probably be on the simple side of that line.